draft/real-world-phoenix-lets-deploy-using-kubernetes.md

```diff
@@ -0,0 +1,643 @@
+If you don't have a project yet, you should create it on gitlab.com and add the
+ID here.
+
+```
+data "gitlab_project" "project-x" {
+  id = [your-project-id]
+}
+```
+
+Then we can use this reference to create the settings in our project:
+
+```
+resource "gitlab_project_cluster" "gitlab-kubernetes" {
+  project                       = data.gitlab_project.project-xid
+  name                          = "my-awesome-cluster"
+  domain                        = "[mydomain.com]"
+  enabled                       = true
+  kubernetes_api_url            = digitalocean_kubernetes_cluster.cl01.endpoint
+  kubernetes_token              = digitalocean_kubernetes_cluster.cl01.kube_config[0].token
+  kubernetes_ca_cert            = base64decode(digitalocean_kubernetes_cluster.cl01.kube_config[0].cluster_ca_certificate)
+  kubernetes_namespace          = ""
+  kubernetes_authorization_type = "rbac"
+  environment_scope             = "*"
+}
+```
+
+And lastly we are not going the skip the database, as that is too easy to do.
+You could potentially setup a database persistent volume in kubernetes and have
+db_pods spinup, but I think that it is much easier to have the
+database as a separate service outside of kubernetes. The digitalocean managed
+database service is a great option. You pay a little extra, but it takes care of
+backups etc. It is just one of those things that you don't want to worry about,
+right?
+
+Ofcourse, we can easily add the creation of a database cluster to our terraform
+setup:
+```json
+resource "digitalocean_database_cluster" "postgres-db" {
+  name       = "postgres-db-cluster"
+  engine     = "pg"
+  version    = "11"
+  size       = "db-s-1vcpu-1gb"
+  region     = "ams3"
+  node_count = 1
+}
+```
+
+And while we are at it, let's create the database and firewall settings as well.
+
+```json
+resource "digitalocean_database_db" "test-prod" {
+  cluster_id = digitalocean_database_cluster.postgres-db.id
+  name       = "test_prod"
+}
+
+resource "digitalocean_database_firewall" "db-fw" {
+  cluster_id = digitalocean_database_cluster.postgres-db.id
+
+  rule {
+    type  = "k8s"
+    value = digitalocean_kubernetes_cluster.cl01.id
+  }
+}
+```
+
+This concludes our cluster setup. To get our app deployed we'll have to add some
+setup to our project. We'll need to dockerize our project, add kubctl deploy
+config files and add a gitlab-ci.yml that will trigger the gitlab-ci pipeline.
+
+## A project to deploy
+
+Now we have our kubernetes cluster up-and-running it is time to see how we would
+deploy our app to the cluster.
+
+I have a git tag prepared for you to use from my real_world_phoenix project. You
+can clone that checkout to use to deploy to the cluster. It has the necessary
+files to trigger deployment which I'll explain next.
+
+Go ahead and clone the project from the tag I created:
+```
+git clone git@gitlab.com:drumusician/real_world_phoenix.git --branch kubernetes-deploy
```

Comment:
Should this still be pushed?

```zsh
$ git clone git@gitlab.com:drumusician/real_world_phoenix.git --branch kubernetes-deploy
Cloning into 'real_world_phoenix'...
remote:
remote: ========================================================================
remote:
remote: You are not allowed to download code from this project.
remote:
remote: ========================================================================
remote:
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Reply:
ah maybe, I'll check

Reply:
@rhynix does this url work: https://gitlab.com/drumusician/real_world_phoenix.git